### PR TITLE
feat: data source from JSON array data

### PR DESCRIPTION
**Pull request by Lilit0x**
Closes: #2218

## How to Test
`SELECT * FROM read_arjson('path/to/file/or/urls', max_size, <creds>)`

### CODE REVIEW

**Review by tychoish — COMMENTED** [review]
I'm super excited about this! 

I wonder if implementing the object store's file type ends up having you write more boilerplate than is needed. 

I had great luck recently using the `StreamingTable` infrastructure in the bson functionality, though I don't know if it makes a lot of sense (also your code supports gzipped inputs and the BSON code does not.)

File: crates/sqlbuiltins/src/functions/table/arr_json/exec.rs [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        let batch_size = self.batch_size;
+        let file_compression_type = self.file_compression_type.to_owned();
+        Ok(Box::pin(async move {
+            let r = store.get(file_meta.location()).await?;
+            let stream = stream! {
+                 match r.payload {
+                GetResultPayload::File(file, _) => {
+                    let decoder = file_compression_type.convert_read(file)?;
+                    let mut reader = BufReader::new(decoder);
+                    let values: Value = serde_json::from_reader(&mut reader)
```
Comment (universalmind303) [inline crates/sqlbuiltins/src/functions/table/arr_json/exec.rs:241]: I'm partial to using using simd-json here. It's generally faster out of the box for reads, and it has a [known-key](https://github.com/simd-lite/simd-json#known-key) feature that makes it even more performant when reading with a known set of keys. I also had much success using it as the base for the json readers in polars

**Comment by Lilit0x** [comment]
> I wonder if implementing the object store's file type ends up having you write more boilerplate than is needed.
You mean the `FileFormat`? Yes, it did. Especially in the `create_physical_plan` function because of the execution. I tried to look for a way around that, but I couldn't since the default `JsonFormat` is new line delimited which gives more credibility to @universalmind303's suggestion to making the change upstream. For now though, I think it is okay like this, albeit, I'll still have to write more for the `create_writer_physical_plan`. 
The other option I had was to create a different type similar to `FileType` and then implement all the traits that datafusion's `FileType` did, but that would have been extremely longer and probably unnecessary. 

> I had great luck recently using the `StreamingTable` infrastructure in the bson functionality, though I don't know if it makes a lot of sense (also your code supports gzipped inputs and the BSON code does not.)
Hmm, I looked through it, I can't seem to see where and how it will fit this use case. Maybe I am missing something.

File: crates/sqlbuiltins/src/functions/table/arr_json/exec.rs (hunk 1 above)
Comment (Lilit0x) [reply crates/sqlbuiltins/src/functions/table/arr_json/exec.rs:241]: This is interesting. And it is compatible with `serde` out of the box. 
Thank youuu!.

**Comment by tychoish** [comment]
I think given that df419aa51d31d99f0d95eb2d2a4b788a3c010ae7 has landed, we should let this one drop. 

I think there's a lot of merit to this approach, and I'm super thankful that you did this work for us to learn from, and sorry that we didn't use the code directly.

**Pull request closed without merging**